lib/rails/decorators/decorator.rb

```diff
@@ -80,7 +62,15 @@ def self.decorate(*targets, &module_definition)
           end
 
           target.const_set(decorator_name, mod)
-          mod.decorates(target)
+
+          if target.name.to_s.end_with?('Helper')
+            engine = target.name =~ /::/ ? target.name.to_s.split('::')[0..-1].join('::') : nil
+            controller = [engine, 'ApplicationController'].join('::').constantize
+
+            controller.helper(mod)
```

Comment:
@bencrouse in http://api.rubyonrails.org/classes/ActionController/Helpers.html, it states that "By default, each controller will include all helpers.", unless `config.action_controller.include_all_helpers` is set to `false`. We can assume that most modern applications won't have that configured, or we could perform an additional check and throw an error if that setting is `false`.

Decorating helpers within engines have slightly different rules. Instead of adding the helper to your main app's `ApplicationController`, you'd want to add the helper to the engine's `ApplicationController` so it's available to all the code in that engine. That's why we attempt to pull the engine name out of the target's constant name. If the helper you're decorating is within an engine, we will add the helper to that engine's `ApplicationController`.